File: backend/app/schemas/metrics.py
"""Read schemas for the analytics metrics endpoints.

These serve dbt-built marts (the `analytics` schema, see transformation/dbt).
Granularity values match the frontend's `Granularity` type so query params map
one-to-one onto the chart filters.

DORA-aligned primitives currently exposed:
  - deployment frequency   → published GitHub Releases
  - pull-request activity  → throughput (opened / merged / closed)
  - cycle time             → lead-time-for-changes proxy (PR open → merge)
  - review latency         → lead-time breakdown / review flow
  - change failure         → change-fail-rate proxy (revert-titled merges)

Additional base primitives:
  - review comments        → GitHub PR review-comment throughput
  - workflow runs          → GitHub Actions run activity
  - tickets / comments / description edits / projects
    (normalized across trackers; source is a mart dimension)
"""

from datetime import date
from typing import Literal

from pydantic import BaseModel

Granularity = Literal["daily", "weekly", "monthly"]

# `org` reads tenant-wide marts; `me` filters to the caller's linked GitHub
# identity (resolved server-side — clients never pass an author).
MetricScope = Literal["org", "me"]


class PullRequestActivityPoint(BaseModel):
    """PR activity bucketed to one period (day/week/month start)."""

    period_start: date
    opened: int
    merged: int
    closed: int


class PullRequestActivityResponse(BaseModel):
    granularity: Granularity
    points: list[PullRequestActivityPoint]


class DeploymentFrequencyPoint(BaseModel):
    """GitHub Release deployment counts bucketed to one period."""

    period_start: date
    releases_published: int
    production_releases: int


class DeploymentFrequencyResponse(BaseModel):
    granularity: Granularity
    points: list[DeploymentFrequencyPoint]


class CycleTimePoint(BaseModel):
    """PR cycle-time (lead-time proxy) bucketed to one period."""

    period_start: date
    prs_merged: int
    median_hours: float
    avg_hours: float
    p90_hours: float


class CycleTimeResponse(BaseModel):
    granularity: Granularity
    points: list[CycleTimePoint]


class ReviewLatencyPoint(BaseModel):
    """Review-latency primitives bucketed to one period."""

    period_start: date
    prs_first_reviewed: int
    median_hours_to_first_review: float | None
    reviews_submitted: int


class ReviewLatencyResponse(BaseModel):
    granularity: Granularity
    points: list[ReviewLatencyPoint]


class ChangeFailurePoint(BaseModel):
    """Change-failure proxy (revert rate) bucketed to one period."""

    period_start: date
    prs_merged: int
    prs_reverted: int
    change_failure_rate: float


class ChangeFailureResponse(BaseModel):
    granularity: Granularity
    points: list[ChangeFailurePoint]


class ReviewCommentsPoint(BaseModel):
    """GitHub PR review-comment throughput bucketed to one period."""

    period_start: date
    review_comments_created: int


class ReviewCommentsResponse(BaseModel):
    granularity: Granularity
    points: list[ReviewCommentsPoint]


class WorkflowRunsPoint(BaseModel):
    """GitHub Actions workflow-run activity bucketed to one period."""

    period_start: date
    runs_started: int
    runs_completed: int
    runs_success: int
    runs_failure: int


class WorkflowRunsResponse(BaseModel):
    granularity: Granularity
    points: list[WorkflowRunsPoint]


class TicketActivityPoint(BaseModel):
    """Ticket activity across issue trackers, bucketed to one period."""

    period_start: date
    tickets_created: int
    tickets_completed: int
    tickets_canceled: int


class TicketActivityResponse(BaseModel):
    granularity: Granularity
    points: list[TicketActivityPoint]


class TicketCommentsPoint(BaseModel):
    """Ticket-comment throughput across issue trackers, bucketed to one period."""

    period_start: date
    comments_created: int


class TicketCommentsResponse(BaseModel):
    granularity: Granularity
    points: list[TicketCommentsPoint]


class ProjectActivityPoint(BaseModel):
    """Project activity across project trackers, bucketed to one period."""

    period_start: date
    projects_created: int
    projects_completed: int
    projects_canceled: int


class ProjectActivityResponse(BaseModel):
    granularity: Granularity
    points: list[ProjectActivityPoint]


class TicketDescriptionEditsPoint(BaseModel):
    """Ticket description-edit counts across issue trackers."""

    period_start: date
    description_edits: int


class TicketDescriptionEditsResponse(BaseModel):
    granularity: Granularity
    points: list[TicketDescriptionEditsPoint]

File: frontend/src/lib/metrics.ts
// Shapes mirror the backend analytics metrics API (app/schemas/metrics.py),
// which serves the dbt-built DORA primitive marts (transformation/dbt). Used by
// the dashboard chart components (home page, My metrics).

import { toIsoDate, type Granularity } from "@/components/charts";
import { authedGet } from "@/lib/api";

/**
 * `org` reads tenant-wide series; `me` restricts PR-based metrics to the
 * signed-in user's authored work (GitHub login resolved server-side).
 */
export type MetricScope = "org" | "me";

export type PullRequestActivityPoint = {
  period_start: string;
  opened: number;
  merged: number;
  closed: number;
};

export type PullRequestActivityResponse = {
  granularity: Granularity;
  points: PullRequestActivityPoint[];
};

export type CycleTimePoint = {
  period_start: string;
  prs_merged: number;
  median_hours: number;
  avg_hours: number;
  p90_hours: number;
};

export type CycleTimeResponse = {
  granularity: Granularity;
  points: CycleTimePoint[];
};

export type ReviewLatencyPoint = {
  period_start: string;
  prs_first_reviewed: number;
  median_hours_to_first_review: number | null;
  reviews_submitted: number;
};

export type ReviewLatencyResponse = {
  granularity: Granularity;
  points: ReviewLatencyPoint[];
};

export type ChangeFailurePoint = {
  period_start: string;
  prs_merged: number;
  prs_reverted: number;
  change_failure_rate: number;
};

export type ChangeFailureResponse = {
  granularity: Granularity;
  points: ChangeFailurePoint[];
};

export type DeploymentFrequencyPoint = {
  period_start: string;
  releases_published: number;
  production_releases: number;
};

export type DeploymentFrequencyResponse = {
  granularity: Granularity;
  points: DeploymentFrequencyPoint[];
};

export type ReviewCommentsPoint = {
  period_start: string;
  review_comments_created: number;
};

export type ReviewCommentsResponse = {
  granularity: Granularity;
  points: ReviewCommentsPoint[];
};

export type WorkflowRunsPoint = {
  period_start: string;
  runs_started: number;
  runs_completed: number;
  runs_success: number;
  runs_failure: number;
};

export type WorkflowRunsResponse = {
  granularity: Granularity;
  points: WorkflowRunsPoint[];
};

export type TicketActivityPoint = {
  period_start: string;
  tickets_created: number;
  tickets_completed: number;
  tickets_canceled: number;
};

export type TicketActivityResponse = {
  granularity: Granularity;
  points: TicketActivityPoint[];
};

export type TicketCommentsPoint = {
  period_start: string;
  comments_created: number;
};

export type TicketCommentsResponse = {
  granularity: Granularity;
  points: TicketCommentsPoint[];
};

export type ProjectActivityPoint = {
  period_start: string;
  projects_created: number;
  projects_completed: number;
  projects_canceled: number;
};

export type ProjectActivityResponse = {
  granularity: Granularity;
  points: ProjectActivityPoint[];
};

export type TicketDescriptionEditsPoint = {
  period_start: string;
  description_edits: number;
};

export type TicketDescriptionEditsResponse = {
  granularity: Granularity;
  points: TicketDescriptionEditsPoint[];
};

type MetricRange = { granularity: Granularity; start: Date; end: Date };
type ScopedMetricRange = MetricRange & { scope?: MetricScope };

function rangeParams(options: ScopedMetricRange): URLSearchParams {
  const params = new URLSearchParams({
    granularity: options.granularity,
    start: toIsoDate(options.start),
    end: toIsoDate(options.end),
  });
  if (options.scope && options.scope !== "org") params.set("scope", options.scope);
  return params;
}

export function getPullRequestActivity(
  token: string,
  tenantId: string,
  options: ScopedMetricRange,
): Promise<PullRequestActivityResponse> {
  const params = rangeParams(options);
  return authedGet<PullRequestActivityResponse>(
    `/api/v1/tenants/${tenantId}/metrics/pull-requests?${params}`,
    token,
  );
}

export function getCycleTime(
  token: string,
  tenantId: string,
  options: ScopedMetricRange,
): Promise<CycleTimeResponse> {
  const params = rangeParams(options);
  return authedGet<CycleTimeResponse>(
    `/api/v1/tenants/${tenantId}/metrics/cycle-time?${params}`,
    token,
  );
}

export function getReviewLatency(
  token: string,
  tenantId: string,
  options: ScopedMetricRange,
): Promise<ReviewLatencyResponse> {
  const params = rangeParams(options);
  return authedGet<ReviewLatencyResponse>(
    `/api/v1/tenants/${tenantId}/metrics/review-latency?${params}`,
    token,
  );
}

export function getChangeFailure(
  token: string,
  tenantId: string,
  options: ScopedMetricRange,
): Promise<ChangeFailureResponse> {
  const params = rangeParams(options);
  return authedGet<ChangeFailureResponse>(
    `/api/v1/tenants/${tenantId}/metrics/change-failure?${params}`,
    token,
  );
}

export function getDeploymentFrequency(
  token: string,
  tenantId: string,
  options: MetricRange,
): Promise<DeploymentFrequencyResponse> {
  const params = rangeParams(options);
  return authedGet<DeploymentFrequencyResponse>(
    `/api/v1/tenants/${tenantId}/metrics/deployment-frequency?${params}`,
    token,
  );
}

export function getReviewComments(
  token: string,
  tenantId: string,
  options: MetricRange,
): Promise<ReviewCommentsResponse> {
  const params = rangeParams(options);
  return authedGet<ReviewCommentsResponse>(
    `/api/v1/tenants/${tenantId}/metrics/review-comments?${params}`,
    token,
  );
}

export function getWorkflowRuns(
  token: string,
  tenantId: string,
  options: MetricRange,
): Promise<WorkflowRunsResponse> {
  const params = rangeParams(options);
  return authedGet<WorkflowRunsResponse>(
    `/api/v1/tenants/${tenantId}/metrics/workflow-runs?${params}`,
    token,
  );
}

export function getTicketActivity(
  token: string,
  tenantId: string,
  options: MetricRange,
): Promise<TicketActivityResponse> {
  const params = rangeParams(options);
  return authedGet<TicketActivityResponse>(
    `/api/v1/tenants/${tenantId}/metrics/tickets?${params}`,
    token,
  );
}

export function getTicketComments(
  token: string,
  tenantId: string,
  options: MetricRange,
): Promise<TicketCommentsResponse> {
  const params = rangeParams(options);
  return authedGet<TicketCommentsResponse>(
    `/api/v1/tenants/${tenantId}/metrics/ticket-comments?${params}`,
    token,
  );
}

export function getProjectActivity(
  token: string,
  tenantId: string,
  options: MetricRange,
): Promise<ProjectActivityResponse> {
  const params = rangeParams(options);
  return authedGet<ProjectActivityResponse>(
    `/api/v1/tenants/${tenantId}/metrics/projects?${params}`,
    token,
  );
}

export function getTicketDescriptionEdits(
  token: string,
  tenantId: string,
  options: MetricRange,
): Promise<TicketDescriptionEditsResponse> {
  const params = rangeParams(options);
  return authedGet<TicketDescriptionEditsResponse>(
    `/api/v1/tenants/${tenantId}/metrics/ticket-description-edits?${params}`,
    token,
  );
}
